src/cpu/pred/bpred_unit.cc

```diff
@@ -123,246 +94,377 @@ BPredUnit::drainSanityCheck() const
+                                            pc.instAddr(),
+                                    hist->indirectHistory);
+
+        if (itarget) {
+            // Indirect predictor hit
+            ++stats.indirectHits;
+            hist->targetProvider = TargetProvider::Indirect;
+            set(hist->target, *itarget);
+
+            DPRINTF(Branch,
+                    "[tid:%i, sn:%llu] Instruction %s predicted "
+                    "indirect target is %s\n",
+                    tid, seqNum, pc, *hist->target);
+        } else {
+            ++stats.indirectMisses;
+            DPRINTF(Branch,
+                    "[tid:%i, sn:%llu] PC:%#x no indirect target\n",
+                    tid, seqNum, pc.instAddr());
+        }
     }
 
-    predHist[tid].push_front(predict_record);
 
-    DPRINTF(Branch,
-            "[tid:%i] [sn:%llu] History entry added. "
-            "predHist.size(): %i\n",
-            tid, seqNum, predHist[tid].size());
+    /** ----------------------------------------------
+     * Fallthrough
+     * -----------------------------------------------
+     * All the target predictors did their job.
+     * If there is no target its either not taken or
+     * a BTB miss. In that case we just fallthrough.
+     * */
+    if (hist->targetProvider == TargetProvider::NoTarget) {
+        set(hist->target, pc);
+        inst->advancePC(*hist->target);
+        hist->predTaken = false;
+    }
+    stats.targetProvider[tid][hist->targetProvider]++;
+
+    // The actual prediction is done.
+    // For now the BPU assume its correct. The update
+    // functions will correct the branch if needed.
+    // If prediction and actual direction are the same
+    // at commit the prediction was correct.
+    hist->actuallyTaken = hist->predTaken;
+    set(pc, *hist->target);
+
+    DPRINTF(Branch, "%s(tid:%i, sn:%i, PC:%#x, %s) -> taken:%i, target:%s "
+            "provider:%s\n", __func__, tid, seqNum, hist->pc,
+            toString(brType), hist->predTaken, *hist->target,
+            enums::TargetProviderStrings[hist->targetProvider]);
+
+
+    /** ----------------------------------------------
+     * Speculative history update
+     * -----------------------------------------------
+     * Now that the prediction is done the predictor
+     * may update its histories speculative. (local
+     * and global path). A later squash will revert
+     * the history update if needed.
+     * The actual prediction tables will updated once
+     * we know the correct direction.
+     **/
+    updateHistories(tid, hist->pc, hist->uncond, hist->predTaken,
+                    hist->target->instAddr(), hist->bpHistory);
+
+
+    if (iPred) {
+        // Update the indirect predictor with the direction prediction
+        iPred->update(tid, seqNum, hist->pc, false, hist->predTaken,
+                      *hist->target, brType, hist->indirectHistory);
+    }
 
-    return pred_taken;
+    // dump();
```

Comment:
That was intentionally kept from before to let users know that this function exists.
Should I remove?

Reply:
Personally I'd either remove it, or leave it uncommented and guarded with a debug flag (if that's possible). A developer should know about its existence from the declaration/definition.

But it would be interesting to hear other opinions on this. @andysan do you have an opinion?

Reply:
I would vote to remove it. These types of comments have a tendency to go stale and it the purpose of this function should be pretty obvious from the header file.